### DISCUSSION TITLE
PCIe Root Port support in Linux-- connecting a device to the Icicle Kit

### DISCUSSION THREAD

**Discussion by mcmordie** [polarfire-soc/polarfire-soc-documentation#40] [Linux drivers for PolarFire Soc]
When I boot the mpfs-dev-cli image I am able to run lspci and see the following:

`root@icicle-kit-es:~# lspci`
`00:00.0 Unclassified device [0006]: Actel Device 1556`

However I do not see the PCIe device which I have attached to the board via J1.  Should this automatically appear or is there something else I need to enable to get this working?  Is there further documentation on the Linux support for the PCIe Root Port?  The documentation I was able to find so far is [here](https://onlinedocs.microchip.com/pr/GUID-E54C85EF-C33B-4E1E-A006-276B04B0FF9B-en-US-2/index.html?GUID-A9064EC2-C497-415F-87B4-22BB3108D465).  This does not talk much about Linux software support.

**Comment by mcmordie** [comment]
Just plugged in a standard Samsung NVMe via an adapter; it works fine:

`00:00.0 Unclassified device [0006]: Actel Device 1556`
`01:00.0 Non-Volatile memory controller: Samsung Electronics Co Ltd NVMe SSD Controller PM9A1/PM9A3/980PRO`

So it looks like the issue I am experiencing is an incompatibility between the PCI Root Port and the device I am trying to use (an M.2 co-processor card).  Any thoughts on why a particular PCIe device would not be detected, or what tools could provide more information?

Is the following !!! Invalid class message significant?

```
root@icicle-kit-es:~# lspci -vvvv
00:00.0 Unclassified device [0006]: Actel Device 1556 (prog-if 04)
        !!! Invalid class 0006 for header type 01
        Control: I/O- Mem- BusMaster- SpecCycle- MemWINV- VGASnoop- ParErr- Stepping- SERR- FastB2B- DisINTx-
        Status: Cap+ 66MHz- UDF- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- >SERR- <PERR- INTx-
        Interrupt: pin D routed to IRQ 0
        Region 0: Memory at <unassigned> (64-bit, prefetchable) [disabled]
        Bus: primary=00, secondary=01, subordinate=01, sec-latency=0
        I/O behind bridge: 0000f000-00000fff [disabled]
        Memory behind bridge: 08000000-081fffff [size=2M]
        Prefetchable memory behind bridge: 0000000008200000-00000000083fffff [size=2M]
        Secondary status: 66MHz- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- <SERR- <PERR-
        BridgeCtl: Parity- SERR+ NoISA- VGA- VGA16- MAbort- >Reset- FastB2B-
                PriDiscTmr- SecDiscTmr- DiscTmrStat- DiscTmrSERREn-
        Capabilities: [80] Express (v2) Root Port (Slot+), MSI 00
                DevCap: MaxPayload 256 bytes, PhantFunc 0
                        ExtTag+ RBE+
                DevCtl: CorrErr- NonFatalErr- FatalErr- UnsupReq-
                        RlxdOrd+ ExtTag+ PhantFunc- AuxPwr- NoSnoop+
                        MaxPayload 128 bytes, MaxReadReq 512 bytes
                DevSta: CorrErr- NonFatalErr+ FatalErr- UnsupReq- AuxPwr- TransPend-
                LnkCap: Port #1, Speed 5GT/s, Width x4, ASPM L0s, Exit Latency L0s <128ns
                        ClockPM- Surprise- LLActRep- BwNot+ ASPMOptComp+
                LnkCtl: ASPM Disabled; RCB 64 bytes, Disabled- CommClk-
                        ExtSynch- ClockPM- AutWidDis- BWInt- AutBWInt-
                LnkSta: Speed 2.5GT/s (downgraded), Width x0 (downgraded)
                        TrErr- Train- SlotClk+ DLActive- BWMgmt- ABWMgmt-
                SltCap: AttnBtn+ PwrCtrl+ MRL+ AttnInd+ PwrInd+ HotPlug+ Surprise+
                        Slot #0, PowerLimit 0.000W; Interlock- NoCompl-
                SltCtl: Enable: AttnBtn- PwrFlt- MRL- PresDet- CmdCplt- HPIrq- LinkChg-
                        Control: AttnInd Off, PwrInd Off, Power- Interlock-
                SltSta: Status: AttnBtn- PowerFlt- MRL- CmdCplt- PresDet- Interlock-
                        Changed: MRL- PresDet- LinkState-
                RootCap: CRSVisible+
                RootCtl: ErrCorrectable- ErrNon-Fatal- ErrFatal- PMEIntEna- CRSVisible+
                RootSta: PME ReqID 0000, PMEStatus- PMEPending-
                DevCap2: Completion Timeout: Range ABCD, TimeoutDis+ NROPrPrP- LTR+
                         10BitTagComp- 10BitTagReq- OBFF Not Supported, ExtFmt+ EETLPPrefix-
                         EmergencyPowerReduction Not Supported, EmergencyPowerReductionInit-
                         FRS- LN System CLS Not Supported, TPHComp- ExtTPHComp- ARIFwd-
                         AtomicOpsCap: Routing- 32bit- 64bit- 128bitCAS-
                DevCtl2: Completion Timeout: 50us to 50ms, TimeoutDis- LTR+ OBFF Disabled, ARIFwd-
                         AtomicOpsCtl: ReqEn- EgressBlck-
                LnkCap2: Supported Link Speeds: 2.5-5GT/s, Crosslink- Retimer- 2Retimers- DRS-
                LnkCtl2: Target Link Speed: 5GT/s, EnterCompliance- SpeedDis-
                         Transmit Margin: Normal Operating Range, EnterModifiedCompliance- ComplianceSOS-
                         Compliance De-emphasis: -6dB
                LnkSta2: Current De-emphasis Level: -6dB, EqualizationComplete- EqualizationPhase1-
                         EqualizationPhase2- EqualizationPhase3- LinkEqualizationRequest-
                         Retimer- 2Retimers- CrosslinkRes: unsupported
        Capabilities: [d0] MSI-X: Enable- Count=1 Masked-
                Vector table: BAR=0 offset=00000000
                PBA: BAR=0 offset=00000000
        Capabilities: [e0] MSI: Enable- Count=1/8 Maskable- 64bit+
                Address: 0000000000000000  Data: 0000
        Capabilities: [f8] Power Management version 3
                Flags: PMEClk- DSI- D1+ D2+ AuxCurrent=0mA PME(D0+,D1+,D2+,D3hot+,D3cold+)
                Status: D0 NoSoftRst+ PME-Enable- DSel=0 DScale=0 PME-
        Capabilities: [100 v1] Vendor Specific Information: ID=1556 Rev=1 Len=008 <?>
        Capabilities: [200 v1] Advanced Error Reporting
                UESta:  DLP- SDES- TLP- FCP- CmpltTO+ CmpltAbrt- UnxCmplt- RxOF- MalfTLP- ECRC- UnsupReq- ACSViol-
                UEMsk:  DLP- SDES- TLP- FCP- CmpltTO- CmpltAbrt- UnxCmplt- RxOF- MalfTLP- ECRC- UnsupReq- ACSViol-
                UESvrt: DLP+ SDES- TLP- FCP+ CmpltTO- CmpltAbrt- UnxCmplt- RxOF- MalfTLP+ ECRC- UnsupReq- ACSViol-
                CESta:  RxErr- BadTLP- BadDLLP- Rollover- Timeout- AdvNonFatalErr-
                CEMsk:  RxErr- BadTLP- BadDLLP- Rollover- Timeout- AdvNonFatalErr+
                AERCap: First Error Pointer: 0e, ECRCGenCap+ ECRCGenEn- ECRCChkCap+ ECRCChkEn-
                        MultHdrRecCap- MultHdrRecEn- TLPPfxPres- HdrLogCap-
                HeaderLog: 00000001 0000000c 000000e0 000000e0
                RootCmd: CERptEn- NFERptEn- FERptEn-
                RootSta: CERcvd- MultCERcvd- UERcvd- MultUERcvd-
                         FirstFatal- NonFatalMsg- FatalMsg- IntMsg 0
                ErrorSrc: ERR_COR: 0000 ERR_FATAL/NONFATAL: 0000
```

The device I am trying to attach has the following lspci verbose info on another platform:

```
01:00.0 Co-processor: Device 1e60:2864 (rev 01)
        Subsystem: Device 1e60:2864
        Control: I/O- Mem+ BusMaster+ SpecCycle- MemWINV- VGASnoop- ParErr+ Stepping- SERR+ FastB2B- DisINTx-
        Status: Cap+ 66MHz- UDF- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- >SERR- <PERR- INTx-
        Latency: 0, Cache Line Size: 64 bytes
        Interrupt: pin A routed to IRQ 20
        Region 0: Memory at 01100000 (64-bit, prefetchable) [size=16K]
        Region 2: Memory at 01108000 (64-bit, prefetchable) [size=4K]
        Region 4: Memory at 01104000 (64-bit, prefetchable) [size=16K]
        Capabilities: [80] Express (v2) Endpoint, MSI 00
                DevCap: MaxPayload 256 bytes, PhantFunc 0, Latency L0s <64ns, L1 unlimited
                        ExtTag+ AttnBtn- AttnInd- PwrInd- RBE+ FLReset+
                DevCtl: Report errors: Correctable- Non-Fatal- Fatal- Unsupported-
                        RlxdOrd+ ExtTag+ PhantFunc- AuxPwr- NoSnoop+ FLReset-
                        MaxPayload 128 bytes, MaxReadReq 512 bytes
                DevSta: CorrErr- UncorrErr- FatalErr- UnsuppReq- AuxPwr- TransPend-
                LnkCap: Port #0, Speed 8GT/s, Width x4, ASPM L0s L1, Exit Latency L0s <1us, L1 <2us
                        ClockPM- Surprise- LLActRep- BwNot-
                LnkCtl: ASPM Disabled; RCB 64 bytes Disabled- CommClk+
                        ExtSynch- ClockPM- AutWidDis- BWInt- AutBWInt-
                LnkSta: Speed 5GT/s, Width x1, TrErr- Train- SlotClk+ DLActive- BWMgmt- ABWMgmt-
                DevCap2: Completion Timeout: Not Supported, TimeoutDis+, LTR+, OBFF Not Supported
                DevCtl2: Completion Timeout: 50us to 50ms, TimeoutDis-, LTR-, OBFF Disabled
                LnkCtl2: Target Link Speed: 8GT/s, EnterCompliance- SpeedDis-
                         Transmit Margin: Normal Operating Range, EnterModifiedCompliance- ComplianceSOS-
                         Compliance De-emphasis: -6dB
                LnkSta2: Current De-emphasis Level: -6dB, EqualizationComplete-, EqualizationPhase1-
                         EqualizationPhase2-, EqualizationPhase3-, LinkEqualizationRequest-
        Capabilities: [e0] MSI: Enable- Count=1/1 Maskable- 64bit+
                Address: 000000004ca38000  Data: 0001
        Capabilities: [f8] Power Management version 3
                Flags: PMEClk- DSI- D1- D2- AuxCurrent=0mA PME(D0-,D1-,D2-,D3hot+,D3cold-)
                Status: D3 NoSoftRst+ PME-Enable- DSel=0 DScale=1 PME-
        Capabilities: [100 v1] Vendor Specific Information: ID=1556 Rev=1 Len=008 <?>
        Capabilities: [108 v1] Latency Tolerance Reporting
                Max snoop latency: 0ns
                Max no snoop latency: 0ns
        Capabilities: [110 v1] L1 PM Substates
                L1SubCap: PCI-PM_L1.2+ PCI-PM_L1.1+ ASPM_L1.2+ ASPM_L1.1+ L1_PM_Substates+
                          PortCommonModeRestoreTime=10us PortTPowerOnTime=10us
        Capabilities: [128 v1] Alternative Routing-ID Interpretation (ARI)
                ARICap: MFVC- ACS-, Next Function: 0
                ARICtl: MFVC- ACS-, Function Group: 0
        Capabilities: [200 v2] Advanced Error Reporting
                UESta:  DLP- SDES- TLP- FCP- CmpltTO- CmpltAbrt- UnxCmplt- RxOF- MalfTLP- ECRC- UnsupReq- ACSViol-
                UEMsk:  DLP- SDES- TLP- FCP- CmpltTO- CmpltAbrt- UnxCmplt- RxOF- MalfTLP- ECRC- UnsupReq- ACSViol-
                UESvrt: DLP+ SDES- TLP- FCP+ CmpltTO- CmpltAbrt- UnxCmplt- RxOF+ MalfTLP+ ECRC- UnsupReq- ACSViol-
                CESta:  RxErr- BadTLP- BadDLLP- Rollover- Timeout- NonFatalErr-
                CEMsk:  RxErr- BadTLP- BadDLLP- Rollover- Timeout- NonFatalErr+
                AERCap: First Error Pointer: 00, GenCap+ CGenEn- ChkCap+ ChkEn-
        Capabilities: [300 v1] #19
```

**Comment by sanjeevb-bansal** [comment]
root@icicle-kit-es:~# lspci
00:00.0 Unclassified device [0006]: Actel Device 1556

Above output of lscpi is correct and expected.  Above line is for the PCI Host Bridge, which is an internal controller, not an external PCIe slave device.

A PCI host bridge provides an interconnect between the processor and peripheral components. Through the PCI host bridge, the processor can directly access main memory independent of other PCI bus masters.

The output above '"lspci output" contains the following sections:-

00:00.0 - The bus number, device number, and function number, in that order.
Unclassified device [0006] :  - Device class.  Typically Host bridge for 00:00.0 BDF value
Actel - Device vendor, Type, PCI. [Vendor ID, 1556. Vendor Name PLDA ( Our PCI IP vendor)] 

Hope this explain the lspci o/p you are seeing.

**Reply by mcmordie** [reply]
Thanks for your response.  Can you think of a reason why the M.2 device listed would fail to enumerate on the bus when an NVMe inserted in the same slot works fine?  This M.2 device is PCIe Gen 3 x4.  From what I understand the PCIe Gen 2 x4 slot on the Icicle kit should be 100% compatible, yet the device fails to enumerate.  The device works fine in another platform.

**Reply by sanjeevb-bansal** [reply]
Pls ignore "!!! Invalid class 0006 for header type 01" message as it is benign. It has no impact for your M.2 device not getting enumerated.

I hope you are using below steps to bring-up Linux on Icicle kit. Pls reprogram card with latest design file and latest Linux image, available here:-

https://github.com/polarfire-soc/polarfire-soc-documentation/blob/master/boards/mpfs-icicle-kit-es/updating-icicle-kit/updating-icicle-kit-design-and-linux.md

Just FYI, I sometime noticed, that my Icicle fail to enumerate my PCI device. Though it's rare. Here is what I do :-

1. Try 2 to 3 power cycle of the board, it works magically. OR
2. Switch off the device and reinsert the PCI device with right orientation. Power back again and I notice that lspci list my PCIe device. OR
3. I would give a bit lift to the PCIe card by placing some 3/4" object below it. This helps to make sure correct electrical contacts of the PCIe card pins with the host PCI slot pins. Now try step one above, hopefully it will work. OR
4. If none of above work try with any other PCIe card e.g. a PCIe NIC etc. 

Hope this helps.

Thanks

**Reply by mcmordie** [reply]
@sanjeevb-bansal  Thanks for your response.  For me NVMe devices are detected 100% of the time in the carrier board but I have never seen the device I need to connect.   I believe this responds to all of your suggestions.

Good to know I can ignore the invalid class message.  I will try pulling changes, rebuilding and seeing if any recent commits fix.

**Reply by mcmordie** [reply]
Whoa, I spoke too soon.  Same distro, one reboot later and I am seeing the device.

Any thoughts on why detection would be intermittent?  Is this an electrical / connectivity issue?

**Reply by sanjeevb-bansal** [reply]
Yes, on some rare kits we need to provide a support (a small raise below PCIe card) and might require one extra reboot  based on my own observation on my own kit. It is probably because of horizontal mount of the PCIe card and hooks to tighten it ( unlike Desktop PC)

**Reply by mcmordie** [reply]
So after most of a day rebooting this thing, I found that the PCIe support for my device (a coprocessor M.2 module) worked once out of maybe 100 reboots, whereas an NVMe inserted in the same adapter board worked 100% of the time.  Successive boots did a variety of the following:

- Froze at 'Starting kernel ...'
- Booted normally but failed to detect the card
- Booted and detected the card, but had failed to load the device firrmware
- Booted, detected the card and successfully loaded firmware, but then disappeared from lspci
- Ran once through successfully

Not sure what explains this wide variety of behavior, but various supports under the board and different insertion depths didn't seemed to make a consistent difference.  Besides the rare connectivity issue, are you aware of any issues with the PCIe support on this platform, either at the RTL level or with the kernel support, which could explain this?  Given than the NVMe operated consistently it appears that something other than connectivity must be an issue.

**Comment by hughbreslin** [comment]
Hey @mcmordie any luck with this? One thing it would be interesting to see, when you get a failure to detect your drive, on the HSS console (MMUART0) could you run a DDR mem test, you just need to type `mem` into the console and see the result. See if it passes or fails, hopefully it passes :)

**Reply by mcmordie** [reply]
@hughbreslin no further progress yet.  Here is a summary / reminder of what I see:

1. My M.2 coprocessor card works sporadically when mounted in an M.2 adapter-- the coprocessor M.2 is what I need to support.
2. For comparison, an NVMe mounted in the same adapter always works-- enumerates fine on PCIe and can be mounted and accessed 
3. Sometimes with the the coprocessor installed the system fails to boot after u-boot.  Last thing I see is "Starting kernel..."
4. Supporting the M.2 adapter with a standoff from the icicle board and adjusting the seating depth in the slot varies the behaviour for the coprocessor card,  Eventually with hundreds of reboots I can get a single test where the coprocessor card loads and runs.

When the unit is failing to boot, this is the MEMTEST result:

```
>> memtest
[40.528366] HSS_MemTestDDRFast(): DDR-Lo size is 32 MiB
[40.534876] HSS_PerfCtr_Allocate(): found already allocated perf ctr for >>MemTest(DDR32)<<
[40.544771] HSS_MemTestDDRFast(): DDR-Hi size is 1888 MiB
[40.551455] HSS_PerfCtr_Allocate(): found already allocated perf ctr for >>MemTest(DDR64)<<
[40.561350] HSS_MemTestDevice(): Write Seed Pattern to all cells
[46.386125] HSS_MemTestDevice(): First Pass: Check each location
[52.476681] HSS_MemTestDevice(): Second Pass: Check each location
[58.525174] tinyCLI_MemTest_(): Passed!
```

The card wiggling issue would be suggestive of a hardware issue-- for example out of spec lane length or impedance-- if the NVMe didn't work consistently.  It would be nice to get to a place of understanding the issue better before our own board gets built.

**Reply by hughbreslin** [reply]
Interesting, could you share which card you're using? I haven't gotten any closer to having a direct answer for you unfortunately. PCIe is something we're currently working on so it could be resolved in a future release but the fact that it works sometimes leads me to believe its something outside of what we're currently working on

**Reply by mcmordie** [reply]
@hughbreslin I can provide more info via email-- mcmordie at viionsystems.com
